RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -507,61 +615,78 @@ DAClusterizerInZT_vect::split(const double beta,  track_t &tks, vertex_t & y, do
+    double sq = sin(qsplit);
+    if(cq < 0){ cq=-cq; sq=-sq; } // prefer cq>0 to keep z-ordering
+
     // estimate subcluster positions and weight
-    double p1=0, z1=0, t1=0, w1=0;
-    double p2=0, z2=0, t2=0, w2=0;
+    double p1=0, z1=0, t1=0, wz1=0, wt1=0;
+    double p2=0, z2=0, t2=0, wz2=0, wt2=0;
     for(unsigned int i=0; i<nt; ++i){
       if (tks.Z_sum_[i] > 1.e-100) {
+	double lr = (tks.z_[i]-y.z_[k]) * cq + (tks.t[i] - y.t_[k]) * sq;
 	// winner-takes-all, usually overestimates splitting
-	double tl = tks.z_[i] < y.z_[k] ? 1.: 0.;
+	double tl = lr < 0 ? 1.: 0.;
 	double tr = 1. - tl;
 
-	// soften it, especially at low T
-	/*
-	double arg = ( tks.z_[i] - y.z_[k] ) * sqrt(beta * tks.dz2_[i]); // + std::fabs(tks.t_[i] - y.t_[k])
-	std::cout << arg << std::endl;
-	if(std::fabs(arg) < 20){
+	// soften it, especially at low T	
+	double arg = lr * std::sqrt(beta * ( cq*cq*tks.dz2_[i] + sq*sq*tks.dt2_[i] ) ); 
+	if(std::abs(arg) < 20){
 	  double t = local_exp(-arg);
 	  tl = t/(t+1.);
 	  tr = 1/(t+1.);
 	}
-	*/
 
 	double p = y.pk_[k] * tks.pi_[i] * local_exp(-beta * Eik(tks.z_[i], y.z_[k], tks.dz2_[i], 
                                                                  tks.t_[i], y.t_[k], tks.dt2_[i])) / tks.Z_sum_[i];
-	double w = p*tks.errsum_[i];
-	p1 += p*tl;  z1 += w*tl*tks.z_[i]; t1 += w*tl*tks.t_[i]; w1 += w*tl;
-	p2 += p*tr;  z2 += w*tr*tks.z_[i]; t2 += w*tr*tks.t_[i]; w2 += w*tr;
+	double wz = p*tks.dz2_[i];
+	double wt = p*tks.dt2_[i];
+	p1 += p*tl;  z1 += wz*tl*tks.z_[i]; t1 += wt*tl*tks.t_[i]; wz1 += wz*tl; wt1 += wt*tl;
+	p2 += p*tr;  z2 += wz*tr*tks.z_[i]; t2 += wt*tr*tks.t_[i]; wz2 += wz*tr; wt2 += wt*tr;
       }
     }
 
-    if(w1>0){z1 = z1/w1; t1 = t1/w1;} else {z1=y.z_[k]-epsilonz; t1=y.t_[k]-epsilont;}
-    if(w2>0){z2 = z2/w2; t2 = t2/w2;} else {z2=y.z_[k]+epsilonz; t2=y.t_[k]+epsilont;}
-    
-    // reduce split size if there is not enough room
-    if( ( k   > 0 ) && ( z1 < (0.6*y.z_[k] + 0.4*y.z_[k-1])) ){ 
-      z1 = 0.5*y.z_[k] + 0.5*y.z_[k-1]; 
-      t1 = 0.5*y.t_[k] + 0.5*y.t_[k-1]; 
-    }
-    if( ( k+1 < nv) && ( z2 > (0.6*y.z_[k] + 0.4*y.z_[k+1])) ){ 
-      z2 = 0.5*y.z_[k] + 0.5*y.z_[k+1]; 
-      t2 = 0.5*y.t_[k] + 0.5*y.t_[k+1]; 
+    if(wz1 > 0){z1 /= wz1;} else {z1 = y.z_[k] - epsilonz * cq; cout << "warning, wz1 = " << scientific << wz1 << endl;}
+    if(wt1 > 0){t1 /= wt1;} else {t1 = y.t_[k] - epsilont * sq; cout << "warning, w11 = " << scientific << wt1 << endl;}
+    if(wz2 > 0){z2 /= wz2;} else {z2 = y.z_[k] + epsilonz * cq; cout << "warning, wz2 = " << scientific << wz2 << endl;}
+    if(wt2 > 0){t2 /= wt2;} else {t2 = y.t_[k] + epsilont * sq; cout << "warning, wt2 = " << scientific << wt2 << endl;}
+
+    unsigned int k_min1, k_min2;
+    while(   (find_nearest(z1, t1, y, k_min1, epsilonz, epsilont) && (k_min1 != k))
+	     || (find_nearest(z2,t2, y, k_min2, epsilonz, epsilont) && (k_min2 !=k)) ){
+      z1 = 0.5*( z1 + y.z_[k]);      z1 = 0.5*( t1 + y.z_[k]);
```

Comment:
There is a typo: rightmost instruction in this line should have a t1 on the left hand side

Reply:
Thanks, on it!